[JENKINS-55832] - Introduce Update Center Plugin plugin info provider for pom.xml plugin inputs

oleg-nenashev opened the pull request:
This Patchs an attempt to provide a more efficient and reliable plugin info retrieval implementation for  Custom WAR Packager. Instead of using a Local maven repository which takes a while to download deps and to resolve the type (and causes false-negatives in the repo when there are network issues) , we just pull the plugin data from the update center. It will not work in the case of unreleased plugins, but it it should help to address common user cases like https://github.com/jenkinsci/ci.jenkins.io-runner . Also, build time for that particular repo has dropped from 10min to less than 1min.

Addressed tickets: 

* https://issues.jenkins-ci.org/browse/JENKINS-55832
* https://issues.jenkins-ci.org/browse/JENKINS-56444, https://issues.jenkins-ci.org/browse/JENKINS-57547, https://issues.jenkins-ci.org/browse/JENKINS-56652, #95 


Changes summary:

- [x] Introduce a `PluginInformationProvider` abstraction so that different implementations can be used
- [x] Introduce new implementation for Update Center installation source 
- [x] **Incompatible change**: Update Center Plugin info provider is no a default mode
- [x] Documentation
- [x] Fix pom.xml input demo which is rather broken ATM
- [x] Add more integration tests

oleg-nenashev commented:
Ready for full review and, if the build passes, for integration

oleg-nenashev commented on README.md:
```diff
@@ -205,6 +205,33 @@ will be omitted. Consequently, if the flag is set to `true` and the pom file doe
+    and it is not recommended for the most of cases
+  * `pomUseMavenPluginInfoProvider: true` in `buildSettings` can be set to enable the mode
+
+Before Custom WAR Packager 
```
Stale comment

oleg-nenashev replied:
```suggestion
Before Custom WAR Packager `1.7`, this provider was used by default.
Builds using this version may need an update if they rely on custom update centers or unreleased/blacklisted plugins.
```

oleg-nenashev commented:
Deployed the image as `jenkins/custom-war-packager:pr-104`. Works well for me

varyvol commented on custom-war-packager-lib/src/main/java/io/jenkins/tools/warpackager/lib/util/MavenHelper.java:
```diff
@@ -134,7 +134,8 @@ public void downloadJAR(File buildDir, DependencyInfo dep, String version, File
         List<DependencyInfo> dependencies = new LinkedList<>();
-        int listed = run(buildDir, true, "dependency:list", "-B", "-DoutputFile=" + destination.getAbsolutePath(),  "-DincludeScope=runtime", "-DexcludeClassifiers=tests", "-f", pom.getAbsolutePath(), "-q");
+        LOGGER.log(Level.INFO, "Listing dependencies from POM file: {0}", pom);
+        int listed = run(buildDir, true, "dependency:list", "-B", "-DoutputFile=" + destination.getAbsolutePath(),  "-DincludeScope=runtime", "-DexcludeClassifiers=tests", "-f", pom.getAbsolutePath());
```
What was the reason for the -q here?

oleg-nenashev replied:
Log suppression. This step takes several minutes on integration tests, and it was just a radio silence

oleg-nenashev commented:
> Only a small comment. I have doubts about changing the default behaviour: how common do you think the use case that won't work with the new provider is?

There would be breaking changes in the following cases:

* Plugins from a custom update center are users. If a custom update center with plugins is used, a user will need to set a custom update center variable
* Unreleased plugins are used (ones not vailable in an update center). In such case the previous Maven Repo Plugin info provider should be used

I do not expect both use-cases above to be common in Jenkins community. But, if it used in proprietary flows, an update will be required when updating the CWP version.

olivierdagenais commented on custom-war-packager-lib/src/main/java/io/jenkins/tools/warpackager/lib/model/plugins/PluginInfoProvider.java:
```diff
@@ -0,0 +1,39 @@
+ * The implementations can be used to query various data about plugins.
+ * @see MavenPluginInfoProvider
+ * @see UpdateCenterPluginInfoProvider
+ * @since TODO
```
P2: Don't forget to punch in the version number once you've made your decision.

oleg-nenashev replied:
That's why it is a TODO comment ;)

olivierdagenais commented on README.md:
```diff
@@ -205,6 +205,34 @@ will be omitted. Consequently, if the flag is set to `true` and the pom file doe
+
+  * The mode caches information about plugins in the Maven repo
+  * The mode is not reliable when used outside clean build environments,
+    and it is not recommended for the most of cases
```
P3: Either "for most of the cases" or, more simply, "for most cases"

oleg-nenashev replied:
```suggestion
    and it is not recommended for most of cases
```

olivierdagenais replied:
I'm not a native English speaker, but "most of cases" doesn't sound right.

oleg-nenashev replied:
Sorry, will fix

olivierdagenais commented on README.md:
```diff
@@ -205,6 +205,34 @@ will be omitted. Consequently, if the flag is set to `true` and the pom file doe
+Information is retrieved from Maven repositories
+
+  * The mode caches information about plugins in the Maven repo
+  * The mode is not reliable when used outside clean build environments,
```
P3: Instead of "The mode", I would recommend "This mode" or, even more precisely, "This provider". (there are 3 instances of that)

olivierdagenais commented on custom-war-packager-lib/src/main/java/io/jenkins/tools/warpackager/lib/config/BuildSettings.java:
```diff
@@ -37,6 +42,9 @@
     @CheckForNull
     private JenkinsfileRunnerSettings jenkinsfileRunner;
+    @CheckForNull
+    private String updateCenterUrl;
```
P2: I'm not seeing any integration tests where a value for the new `updateCenterUrl` parameter/setting is provided.  Seeing as there are getters and setters for all the other fields in this class, I'm wondering if there's magic involved (the value will be automatically deserialized without needing to implement a getter/setter pair) or simply an omission.

oleg-nenashev replied:
There is no stable update center I could use for such test. Expoerimental Update Center (https://jenkins.io/doc/developer/publishing/releasing-experimental-updates/) could be used, but it would be matter of time till test assumptions change (e.g. an experimental-only plugin gets released). Setting up a test update center is doable, but it is an overkill IMHO. I can create a follow-up JIRA ticket

oleg-nenashev replied:
I added a smoke test with a TODO comment for now

olivierdagenais replied:
It might be sufficient to use wiremock to set up a simple localhost endpoint that returns some unique JSON payload, just to confirm that the de-serialization is taking place correctly and the custom value can be used.

olivierdagenais commented on custom-war-packager-lib/src/main/java/io/jenkins/tools/warpackager/lib/impl/plugins/UpdateCenterPluginInfoProvider.java:
```diff
@@ -0,0 +1,61 @@
+        JSONObject jsonObj = JSONObject.fromObject(json);
+
+        // UpdateSite.Plugin does not contain gav object, so we process the JSON object on our own here
+        for(Map.Entry<String,JSONObject> e : (Set<Map.Entry<String,JSONObject>>)jsonObj.getJSONObject("plugins").entrySet()) {
```
P2: A bit of refactoring to extract a few unit-testable methods for this parsing would help us write unit tests that would hopefully help me understand better/faster what's going on here and what the key and value pairs are.

After formatting the JSON I got from the update center, it looks like, given a fragment like:
```
        "AnchorChain": {
            "buildDate": "Mar 11, 2012",
(...)
            "gav": "org.jenkins-ci.plugins:AnchorChain:1.0",
(...)
            "wiki": "https://plugins.jenkins.io/AnchorChain"
        },
```

...the key is `AnchorChain`, which appears to be the extracted from the artifactId portion of the `gav` value, and thus we are creating a lookup of plugins by their artifactId and storing their groupId as the associated value.

If I understood that right, I would like to make a suggestion: instead of creating a `HashMap<>`, consider creating a `TreeMap<>(String.CASE_INSENSITIVE_ORDER)`.  This will impact performance somewhat (`O(log(n))` instead of `O(1)`) but then users won't get "plugin not found" if they don't have the exact casing of the artifactId, as, until now, I thought all plugins had lowercased artifactId. (assuming, of course, there don't exist distinct plugins whose names differ only by case)

oleg-nenashev replied:
> If I understood that right, I would like to make a suggestion: instead of creating a HashMap<>, consider creating a TreeMap<>(String.CASE_INSENSITIVE_ORDER). This will impact performance somewhat (O(log(n)) instead of O(1)) but then users won't get "plugin not found" if they don't have the exact casing of the artifactId, as, until now, I thought all plugins had lowercased artifactId. (assuming, of course, there don't exist distinct plugins whose names differ only by case)

I prefer not to touch it, because it may lead to artifact ID resolution issues in some cases. I also plan to use https://github.com/jenkinsci/plugin-installation-manager-tool in this repo eventually, and I would prefer not to touch the case sensitivity until there is a support for it in the underlying tool. CC @stopalopa , It might be an interesting RFE for your project

oleg-nenashev commented on README.md:
```diff
@@ -205,6 +205,34 @@ will be omitted. Consequently, if the flag is set to `true` and the pom file doe
+
+#### Update Center Information provider
+
+The mode was introduced in Custom WAR Packager `1.8`,
```
We do not use semver in Jenkins as a standard, but I agree it makes sense to switch. Some extra cleqnup could be done then

oleg-nenashev replied:
Will switch to Semver in the next release

oleg-nenashev commented on custom-war-packager-lib/src/main/java/io/jenkins/tools/warpackager/lib/impl/plugins/UpdateCenterPluginInfoProvider.java:
```diff
@@ -0,0 +1,61 @@
+
+public class UpdateCenterPluginInfoProvider implements PluginInfoProvider {
+
+    public static final String DEFAULT_JENKINS_UC_URL = "http://updates.jenkins.io/update-center.json";
```
Agreed, will change the default

oleg-nenashev commented on custom-war-packager-lib/src/main/java/io/jenkins/tools/warpackager/lib/config/BuildSettings.java:
```diff
@@ -37,6 +42,9 @@
     private JenkinsfileRunnerSettings jenkinsfileRunner;
+    @CheckForNull
+    private String updateCenterUrl;
+    private boolean pomUseMavenPluginInfoProvider;
```
It is a subject for a bigger rework later. I am not happy about how the settings are organized, and I would like to make a breaking change in the format in one of the next releases

oleg-nenashev commented:
I hope I addressed the feedback from @olivierdagenais . I am going to merge this change so that i can create some downstream pull requests. If anything else pops up in reviews here, I do not plan to release new version till Tuesday. I will be happy to address any feedback in follow-up PRs
